crates/ty_python_semantic/src/semantic_index/member.rs

```diff
@@ -0,0 +1,395 @@
+    /// use [`Self::as_instance_attribute`].
+    pub(super) fn as_instance_attribute_candidate(&self) -> Option<&Name> {
+        match &*self.expression.segments {
+            [MemberSegment::Attribute(name)] => {
+                // The last segment is a symbol, the second is an attribute.
+                Some(name)
+            }
+            _ => None,
+        }
+    }
+
+    /// Return `true` if the place expression has the form `<NAME>.<MEMBER>`,
+    /// indicating that it *may* be an instance attribute if we are in a method context.
+    ///
+    /// This method is internal to the semantic-index submodule.
+    /// It *only* checks that the AST structure of the `Place` is
+    /// correct. It does not check whether the `Place` actually occurred in
+    /// a method context, or whether the `<NAME>` actually refers to the first
+    /// parameter of the method (i.e. `self`). To answer those questions,
+    /// use [`Self::is_instance_attribute`].
+    pub(super) fn is_instance_attribute_candidate(&self) -> bool {
+        self.as_instance_attribute_candidate().is_some()
+    }
+
+    /// Does the place expression have the form `self.{name}` (`self` is the first parameter of the method)?
+    pub(super) fn is_instance_attribute_named(&self, name: &str) -> bool {
+        self.as_instance_attribute().map(Name::as_str) == Some(name)
+    }
+
+    /// Return `Some(<ATTRIBUTE>)` if the place expression is an instance attribute.
+    pub(crate) fn as_instance_attribute(&self) -> Option<&Name> {
+        if self.is_instance_attribute() {
+            debug_assert!(self.as_instance_attribute_candidate().is_some());
+            self.as_instance_attribute_candidate()
+        } else {
+            None
+        }
+    }
+}
+
+impl std::fmt::Display for Member {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Display::fmt(&self.expression, f)
+    }
+}
+
+bitflags! {
+    /// Flags that can be queried to obtain information about a member in a given scope.
+    ///
+    /// See the doc-comment at the top of [`super::use_def`] for explanations of what it
+    /// means for a member to be *bound* as opposed to *declared*.
+    #[derive(Copy, Clone, Debug, Eq, PartialEq)]
+     struct MemberFlags: u8 {
+        const IS_BOUND              = 1 << 0;
+        const IS_DECLARED           = 1 << 1;
+        const IS_INSTANCE_ATTRIBUTE = 1 << 2;
+    }
+}
+
+impl get_size2::GetSize for MemberFlags {}
+
+/// The member expression consists of different segments, e.g. `x.y.z` is represented
+/// as `Attribute(y), Attribute(z)` segments (in that order).
+///
+/// Note: The segments are internally stored in reverse order. This allows constructing
+/// a `MemberExpr` from an ast expression without having to reverse the segments.
+#[derive(Clone, Debug, PartialEq, Eq, get_size2::GetSize, Hash)]
+pub(crate) struct MemberExpr {
+    symbol_name: Name,
+    segments: SmallVec<[MemberSegment; 1]>,
```

Comment:
I think we can optimize this further by: 

* Storing a single `Name` which is the entire path of the member
* In the `SmallVec`: Store the kind of each segment and where it starts as a `u32`. 

This should be sufficient because all we need is a unique key to hash, knowing if it has the form `a.b`, and a `Display` implementation